### ISSUE TITLE
MathJax SVG math renderer broken on Windows Chrome

### ISSUE THREAD

**Issue by trombonekenny** [PrairieLearn/PrairieLearn#6495]
A student reported that when he accidentally set the MathJax Math Renderer to SVG the rendering no longer happens and the raw MathML is posted to the screen (with https://www.prairielearn.org/pl/course_instance/4970/instructor/question/1793638/preview):

![image](https://user-images.githubusercontent.com/10765199/196295748-91abb920-969c-411d-8791-1cd82f7fafae.png)

In Windows 10 Chrome, after changing that to SVG, this error is in the Console and nothing is displayed for the math:

```
TypeError: Failed to execute 'appendChild' on 'Node': parameter 1 is not of type 'Node'.
    at e.append (core.js:1:2475)
    at t.node (core.js:1:7882)
    at e.html (svg.js:1:20057)
    at r.svg (svg.js:1:118040)
    at r.pageElements (svg.js:1:115678)
    at e.documentPageElements (core.js:1:169324)
    at e.addPageElements (core.js:1:168225)
    at e.updateDocument (core.js:1:168019)
    at t.updateDocument (menu.js:1:31573)
    at Object.renderDoc (core.js:1:15620)
```
On page reload, with LocalStorage `MathJax-Menu-Settings` set to `{"renderer":"SVG"}` shows the raw mathML:

![image](https://user-images.githubusercontent.com/10765199/196296072-79dcbd20-22e7-4b4c-bace-d69ec019586c.png)

and these errors on the console:

```
TypeError: Failed to execute 'appendChild' on 'Node': parameter 1 is not of type 'Node'.
    at e.append (core.js:1:2475)
    at t.node (core.js:1:7882)
    at e.html (svg.js:1:20057)
    at r.svg (svg.js:1:118040)
    at r.pageElements (svg.js:1:115678)
    at e.documentPageElements (core.js:1:169324)
    at e.addPageElements (core.js:1:168225)
    at e.updateDocument (core.js:1:168019)
    at t.updateDocument (menu.js:1:31573)
    at Object.renderDoc (core.js:1:15620)
core.js:1 Uncaught (in promise) TypeError: Failed to execute 'appendChild' on 'Node': parameter 1 is not of type 'Node'.
    at e.append (core.js:1:2475)
    at t.node (core.js:1:7882)
    at e.html (svg.js:1:20057)
    at r.svg (svg.js:1:118040)
    at r.pageElements (svg.js:1:115678)
    at e.documentPageElements (core.js:1:169324)
    at e.addPageElements (core.js:1:168225)
    at e.updateDocument (core.js:1:168019)
    at t.updateDocument (menu.js:1:31573)
    at Object.renderDoc (core.js:1:15620)
```

In Slack, @mwest1066 reported SVG rendering was working under MacOS browsers, and @echuber2 said it worked on Linux Firefox.

I feel like there's a bug in how we include mathjax components? Or maybe an upstream bug in their implementation?

This also identified that the solution to fix it means going into the Developer Tools, Application, and removing the LocalStorage key that configures MathJax to non-defaults. When the MathML was displayed above, right clicking it didn't do anything special, so when broken there's no easy way to change back.
- Maybe we need something under `/pl/settings` to let the user view and remove their PL localStorage variables?

**Comment by echuber2** [comment]
It seems like there might be another underlying issue. Even on Windows Chrome, the SVG mode _should_ at least finish the render routine without crashing. The issue with SVG on Win+Chrome in the past was that thin lines would render as <1px thickness (hence invisible) at some screen resolutions, due to an underlying bug in the Windows rendering libraries that Chrome (but not Firefox) relies on.

Did deleting the MathJax preference in the local storage resolve the issue? What happens if the student switches to Windows Firefox temporarily?

**Comment by echuber2** [comment]
I think recent improvements to the ACE editor are also saving user prefs in local storage, so it's probably better not to wipe them all out directly. Maybe the PL user settings page could instead have a viewer of the current local storage and let the user pick an item to remove (but not edit). This isn't a very elegant solution to the Latex issue though, since any users suffering from this sporadically would have to be informed to go find the setting and clear it anyway. Each latex item could have one of the "?" tooltip anchors next to it as some PL elements do, and this could have a diagnostic control on it for clearing the preferences, but that might clutter the appearance of latex overall.

**Comment by trombonekenny** [comment]
I saw the same errors and performance under Windows Firefox.

**Comment by echuber2** [comment]
Hm, that doesn't sound good. Does clearing the local storage indeed fix it for you? Can you try to reproduce the issue using the same user account on the same question, but on a different OS or machine? I'll see if I can replicate in Windows myself.

**Comment by nwalters512** [comment]
I can reproduce this on macOS Chrome. I haven't yet had a chance to look into a root cause or anything, but at least I can reproduce it (and ultimately verify a fix) without a Windows machine.

**Comment by nwalters512** [comment]
GitHub's implementation of MathJax just straight-up disables the "renderer" option! Try it for yourself:

$$ x = 1 $$

They ship sourcemaps, so I can tell how they did it:

```ts
const menu = window.MathJax.startup.document!.menu.menu
const renderer = menu.find('Renderer')
renderer?.disable()
```

This might be a simpler fix, and also more foolproof for end users.

**Comment by echuber2** [comment]
Something like:

```js
try {
  const settings = JSON.parse(localStorage.getItem("MathJax-Menu-Settings"));
  delete settings["renderer"];
  const settingsJson = JSON.stringify(settings);
  if (settingsJson === "{}") {
    localStorage.removeItem("MathJax-Menu-Settings");
  } else {
    localStorage.setItem("MathJax-Menu-Settings", settingsJson);
  }
} catch(e) { }
```

**Comment by trombonekenny** [comment]
It's reproducible outside of PL: https://jsfiddle.net/rLvp0e7m/7/

I'll see if I can find/file the issue with MathJax.

**Comment by trombonekenny** [comment]
https://github.com/mathjax/MathJax/issues/2956

**Comment by echuber2** [comment]
The MathJax maintainer says this will be fixed in v4.0 but that may be a ways off, and it looks like it has many other changes. In the meantime, what is the best solution?

- Set cache to `local`
- Deactivate renderer menu
- Clear local storage selection for renderer
- All of the above

**Comment by echuber2** [comment]
Just to bump this and add another angle: I think for the time being, the simplest thing to do is to change cache to `local`, since we can leave alone the preference setting the user has chosen that way. The CHTML renderer on Win+Chrome was _supposed_ to be working to fix thin lines, but we got a report that the issue was happening again in spite of that for some users. So maybe we need a different fix anyway for Win+Chrome.

After some experimentation, it seems that setting a small CSS transform rotation doesn't really help with thin horizontal/vertical lines (disappointingly). However, making fonts bigger seems to help. Perhaps on Win+Chrome we should set a CSS zoom transform or increase the base text size. (Maybe this could just be scoped to the MathJax rendered element.)
